doc/src/devdocs/types.md

```diff
@@ -315,7 +315,129 @@ Ptr{Void} @0x00007f5998a04370
 
 so `Tuple == Tuple{Vararg{Any}}` is indeed the primary type.
 
-## Introduction to the internal machinery: `jltypes.c`
+## Diagonal types
+
+Consider the type `Tuple{T,T} where T`.
+A method with this signature would look like:
+
+```julia
+f(x::T, y::T) where T = ...
+```
+
+According to the usual interpretation of a `UnionAll` type, this `T` ranges over all
+types, including `Any`, so this type should be equivalent to `Tuple{Any,Any}`.
+However, this interpretation causes some practical problems.
+
+First, a value of `T` needs to be available inside the method definition.
+For a call like `f(1, 1.0)`, it's not clear what `T` should be.
+It could be `Union{Int,Float64}`, or perhaps `Real`.
+Intuitively, we expect the declaration `x::T` to mean `T === typeof(x)`.
+To make sure that invariant holds, we need `typeof(x) === typeof(y) === T` in this method.
+That implies the method should only be called for arguments of the exact same type.
+
+It turns out that being able to dispatch on whether two values have the same type
+is very useful (this is used by the promotion system for example), so we have
+multiple reasons to want a different interpretation of `Tuple{T,T} where T`.
+To make this work we add the following rule to subtyping: if a variable occurs
+more than once in covariant position, it is restricted to ranging over only concrete
+types.
+("Covariant position" means that only `Tuple` and `Union` types occur between an
+occurrence of a variable and the `UnionAll` type that introduces it.)
+Such variables are called "diagonal variables" or "concrete variables".
+
+So for example, `Tuple{T,T} where T` can be seen as
+`Union{Tuple{Int8,Int8}, Tuple{Int16,Int16}, ...}`, where `T` ranges over all
+concrete types.
+This gives rise to some interesting subtyping results.
+For example `Tuple{Real,Real}` is not a subtype of `Tuple{T,T} where T`, because
+it includes some types like `Tuple{Int8,Int16}` where the two elements have
+different types.
+`Tuple{Real,Real}` and `Tuple{T,T} where T` have the non-trivial intersection
+`Tuple{T,T} where T<:Real`.
+However, `Tuple{Real}` *is* a subtype of `Tuple{T} where T`, because in that case
+`T` occurs only once and so is not diagonal.
+
+Next consider a signature like the following:
+
+```julia
+f(a::Array{T}, x::T, y::T) where T = ...
+```
+
+In this case, `T` occurs in invariant position inside `Array{T}`.
+That means whatever type of array is passed unambiguously determines
+the value of `T` --- we say `T` has an *equality constraint* on it.
+Therefore in this case the diagonal rule is not really necessary, since
+the array determines `T` and we can then allow `x` and `y` to be of
+any subtypes of `T`.
+So variables that occur in invariant position are never considered diagonal.
+This choice of behavior is slightly controversial, but your humble documentation
+author (Jeff in this case) likes it.
```

Comment:
I would just delete this last sentence. Either that or explain why it's controversial and then provide a counterargument to that which justifies this behavior. The alternatives seem to be:

1. Do what we're doing now.
2. Change the diagonal behavior above and lose the ability to dispatch on concrete diagonals only.
3. Require `T` here to be concrete and lose the ability to match types with abstract parameters.

What we're doing now is the most expressive and intuitive. I can see how the concrete diagonal behavior might be controversial, but given that, I can't see how this part would be.